### ISSUE TITLE
You can provide objectbox plugin source code?

### ISSUE THREAD

**Issue by sjdlfajdlfe** [objectbox/objectbox-java#560]

**Comment by greenrobot-team** [comment]
I don't know if there are plans to make the code public, right now it is not available.

cc @greenrobot

**Comment by greenrobot-team** [comment]
@HugoGresse The ObjectBox plugin is only used at build time so using it shouldn't cause any issues regarding licensing of your code (but again, depends on your policy of course).

**Comment by HugoGresse** [comment]
Indeed, at build time, which F-Droid does. Every app on the F-Droid FOSS store is built by F-Droid server which prevent the final build to include non-opensource code (or code which was not on the repo itself).

**Comment by LeoColman** [comment]
Hey guys!
Any way I can help on this? My app got blocked in FDroid due to this issue, and my current plan is to move away from object box
But I really like it, so maybe there's something we can do 😁

**Comment by growse** [comment]
It feels a *tiny* bit disingenous to have a GH repo where the `LICENSE.txt` file says "Apache-2.0 license" but the `README.txt` next to it says "Hey, just add this closed-source plugin to your gradle file"...

**Comment by greenrobot** [comment]
> Indeed, at build time, which F-Droid does. Every app on the F-Droid FOSS store is built by F-Droid server

OK, that's actually a good reason to prioritize open sourcing the plugin a bit higher...

Does anybody know if F-droid also checks for binaries?

**Comment by LeoColman** [comment]
@greenrobot 
Yes, binaries are scanned. 
ref: https://gitlab.com/fdroid/fdroidserver/-/merge_requests/1150#note_1107739881

**Comment by kelson42** [comment]
@greenrobot This problem has been raised by F-Droid team and the impact is really high: for good or bad reasons all recent version of Kiwix has been removed from F-Droid because of this problem. It's important for me to:
* understand the rational behind the decision behind making this compilation tool non-free
* get a kind of timeline for the release on open-source.

This is important to me to assess the situation and decide if it makes sense to replace objectbox (we are super happy with btw) or keep it and wait the fix.

**Comment by growse** [comment]
Related: what's the license status of `libobjectbox-jni.so`? Maven (https://mvnrepository.com/artifact/io.objectbox/objectbox-android/3.3.1) refers to the "ObjectBox Binary Licence", but I can't find any reference to what that means.

> This is important to me to assess the situation and decide if it makes sense to replace objectbox (we are super happy with btw) or keep it and wait the fix.

I imagine quite a few people are in the same boat. With new releases blocked on F-Droid, projects are likely finding themselves trying to decide whether to start working on a migration immediately, or wait for developments.

**Comment by kelson42** [comment]
> With new releases blocked on F-Droid

No, it's far worse. All releases relying on this for building have been removed.

**Comment by eighthave** [comment]
F-Droid's policies on free software come from Free Software Foundation, Open Source Initiative, Debian, etc. They have clearly defined what free software is, so we don't need to.  GNU/Linux distros like RedHat, Fedora, Debian, Ubuntu, etc. all build all packages from source, no outside binaries are allowed in the official release process.  The Debian builders don't allow network access even.

An easy way to think about what free software is: can the user customize it using only free software? If the user needs to use non-free software to build the source code, then the user does not have the freedom to modify the software in any way they want to. They cannot change the proprietary build tool or plugin.

**Comment by kelson42** [comment]
@eighthave Your comment is misleading, unfortunately. AFAIK none of the OSI recognized [open source licenses](https://opensource.org/licenses), and FSF one included, put any constraint on the build dependences (dependences needed to build but not included in the software). If we follow your logic, and this is only a counter-example, Thunderbird won't be a FOSS because [it requires Visual Studio (propriatery software) to compile on Windows](https://opensource.org/licenses), this is non-sense.

F-Droid or Debian, as software distributors and software builders, have their own rules, which are perfectly legit. But their rules apply only to their own garden: they don't define what is FOSS or not. To the contrary to the FSF for example, which is a FOSS license publisher (and enforcer).

F-Droid rules are legit for F-Droid and that is good enough. My advice would be that, instead of challenging FOSS developers with this kind of wrong - and slightly insulting - assertion, you better focus about how to communicate around your own rules, think about how to enforce them without putting users and publishers in difficulty within 24 hours...

To conclude, asserting that softwares relying on this - meanwhile - infamous non-free Objectbox plugin, are non-free themselves is false and confusioning. You are the second contributor of F-Droid I read making this mistake and this is not desserving the FOSS movement IMO.

**Comment by growse** [comment]
> (dependences needed to build but not included in the software

The plugin is not just a build dependency. It pulls in https://search.maven.org/artifact/io.objectbox/objectbox-android/3.3.1/aar into the implementation dependencies, which itself then pulls in a closed-source native binary, which is then bundled with the released software.

**Comment by eighthave** [comment]
Thunderbird does not require Windows nor Visual Studio, it does support them as 
an option.  It is free software because you can build it and run it on Debian. 
Indeed it is included in Debian.  Free software is about user freedom, so users 
should be free to use proprietary tools if they want.  The key point is that 
users are not required to use Windows or Visual Studio.  objectbox will not be 
free software as long as it requires users to use proprietary software to build 
or run it.

If you're talking "open source", that' looser definition.  Some people even call 
proprietary software open source if they can read the source code.

**Comment by kelson42** [comment]
> Thunderbird does not require Windows nor Visual Studio, it does support them as an option. It is free software because you can build it and run it on Debian.

Wow, so basically this is because Thunderbird is multi-platform that it is FOSS. This is getting better and better over time! So lets take a Windows only free software: Notepad++ is not FOSS?

**Comment by kelson42** [comment]
> @eighthave @kelson42 
> 
> That's beyond the point. It's not a build-only dependency, it ships a non-foss binary
> 
> https://github.com/objectbox/objectbox-java/blob/556d9652babda6a7aeca64396dc827bea8f35e82/objectbox-java/src/main/java/io/objectbox/internal/NativeLibraryLoader.java
> 
> 
> 
> Specifically this line: https://github.com/objectbox/objectbox-java/blob/556d9652babda6a7aeca64396dc827bea8f35e82/objectbox-java/src/main/java/io/objectbox/internal/NativeLibraryLoader.java#L103
> 
> 
> 
> that needs to be loaded in order to function.

@LeoColman Thx, this is the key point really missing. It's not a build dependence to a proprietary software which is a problem, actually this is a runtime dependence to propriatery software. This is obviously the real problem. It seems this is not clearly stated in this repository readme (or I missed so ething) and the Apache license of this java binding is misleading. We should really have been more careful at the time it has been introduced in Kiwix. The price for this error is high.

**Comment by greenrobot-team** [comment]
Thanks for your input!

Here is an overview of the current licenses that is hopefully helpful:

| Package | License | Source code | Usage |
| - | - | - | - |
| Java libraries | [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0) | [open](https://github.com/objectbox/objectbox-java) | Runtime |
| Gradle plugin | [GNU General Public License, Version 3](https://www.gnu.org/licenses/gpl.txt) | closed | Build time |
| Native libraries | [ObjectBox Binary Licence](https://objectbox.io/0209-ob-binary-license/) | closed | Runtime |

Note that the Gradle plugin adds Java and native libraries to your project dependencies as needed.

[Maven Central search](https://search.maven.org/search?q=g:io.objectbox) is an easy way to see all dependencies and their licenses.

**Comment by LeoColman** [comment]
@greenrobot-team 
I've created pull request #1104 to add the license information to the readme.
I think this table you provided should be placed everywhere license related

**Comment by natanfudge** [comment]
I would like to add, making the plugin open source (not necessarily the proprietary native code) would allow making contributions to it. Tickets such as #1075 and #948 have been open for a long time, and the community could definitely help, and have already offered to help, in solving them.

**Comment by niltonvasques** [comment]
Any update here regard the source code of these binaries? I'm debugging some critical memory leaks in my application, and the dumps are pointing to some internal C++ objbox native library calls, that I unfortunately cannot debug by myself, since the code is not available. BTW we use objbox for many years now, and just today I realized that it is not entirely open source.

**Comment by greenrobot-team** [comment]
@niltonvasques Feel free to open an issue about this here. We also offer professional (commercial) support if you need it.

**Comment by marek22k** [comment]
Is there any news? Some F-Droid app are still not available :-(

**Comment by greenrobot-team** [comment]
@marek22k If there is an update, we will share it. Please only comment if you have something to contribute (e.g. an implementation suggestion, a use case that wasn't mentioned or where the suggested workaround does not work). Otherwise this produces needless work for us (e.g. me checking all new comments for something relevant).

I will hide such comments so it is easier for us and other users to see relevant comments.

If you are interested in this, please thumbs up the first post instead. This helps us track interest.

**Comment by mnalis** [comment]
Any reason why PR #1104 (which clarified quite confusing licensing situation) was closed without comments? Was it mistake (head was deleted?) ?

It was mostly based on @greenrobot-team comment at https://github.com/objectbox/objectbox-java/issues/560#issuecomment-1266858993. Can it be reopened? If there are specific improvements to be made, could it be commented what those are, so it can be ammended?  

In current form, licensing is misleading at the least. It would benefit everyone if legal uncertainty were clarified instead.

**Comment by greenrobot-team** [comment]
@mnalis The PR was closed because we made [updates to the README ourselves](https://github.com/objectbox/objectbox-java/commit/4aa80e33d041ac7a253af50cfc29d47d19c1ba67). It now links to a [license info entry](https://objectbox.io/faq/#license-pricing) on our website that mirrors my comment above. This way we can re-use it all our projects.

**Comment by mnalis** [comment]
@greenrobot few comments:

- I'd suggest to keep such information directly in the README itself. Relying on website resourse has many disadvantages (URL changes and other bitrot happens, licensing information that is valid today might not be valid in 5 years, there is no log/history, offline copies of the code lack the information for verification etc). Keeping it in the git repository `README`  with rest of the code whose situation it describes solves all those problems. You can link from website to that README, if such avoiding duplication of information is preferred by you.
- web version seems to miss most important column from your comment, i.e. `Source code: open/closed`. That significantly reduces its usefulness for interested parties.
- I would like clarification in particular about  `Package: Gradle plugin` / `License: GNU General Public License, Version 3` / `Source code: closed` / `Usage: Build time`. Can you explain how that works exactly (GPLv3 + closed source)?

**Comment by greenrobot-team** [comment]
Thanks for the thoughts regarding inlining vs. linking. We decided to add a link to the website for the reasons stated.

Regarding publishing of the plugin source code: licensing under GPL does not require to also publish the code. Only if GPL code of another author is modified and used, the changes need to be made publicly available. Anyhow, we plan to publish the plugin at some point, but as the repo currently has some internal stuff and due to lack of time this has still not happened.

**Comment by papjul** [comment]
> licensing under GPL does not require to also publish the code. Only if GPL code of another author is modified and used, the changes need to be made publicly available

I don't know where you got that idea. All users of a program under GPL license are granted the freedom to access the source code, no exception.

Quoting preamble of the license:

> Our General Public Licenses are designed to make sure that you have the freedom to distribute copies of free software (and charge for them if you wish), that **you receive source code or can get it if you want it**, that you **can change the software or use pieces of it in new free programs**, and that you know you can do these things.

> To protect your rights, we need to **prevent others from denying you these rights or asking you to surrender the rights**. Therefore, **you have certain responsibilities if you distribute copies of the software**, or if you modify it: responsibilities to **respect the freedom of others**.

> For example, **if you distribute copies of such a program**, whether gratis or for a fee, you must pass on to the recipients the same freedoms that you received. **You must make sure that they, too, receive or can get the source code. And you must show them these terms so they know their rights.**

> Developers that use the GNU GPL protect your rights with two steps: (1) assert copyright on the software, and (2) offer you **this License giving you legal permission to copy, distribute and/or modify it**.

As you can see, when you distribute a program under GPL license, you have responsibilities. One of them is to ensure all users of your programs can get access to the source code.

Additionally, section 12 says:

> If you cannot convey a covered work so as to satisfy simultaneously your obligations under this License and any other pertinent obligations, then as a consequence you may not convey it at all.

Basically, this means you can't apply the GPL license partially. You must either follow all your obligations (one of them being to let users access the source code) for all your GPL programs, OR cease the distribution of these programs under GPL license (unless your programs contain GPL parts of other authors, in which case, that option is not possible).

**Comment by LeoColman** [comment]
@papjul it's on purpose. Your argument is wrong because the other side doesn't care
#1100 #783

**Comment by papjul** [comment]
None of the issues you linked ever mention "GPL" or "violation".

If you're not part of the ObjectBox team (or so I guess from your previous comments in this issue), it would be wise to not speak in their name about a legal issue.

Contrary to the other cases you mention, they have to provide a clarification. Either it's closed source, or it's GPLv3 AND open source. There is no " it's GPL and source code will be posted later" answer possibility.

**Comment by greenrobot-team** [comment]
The Gradle plugin including the generator source code is now available at https://github.com/objectbox/objectbox-java-generator. Sorry that it took us this long, but as mentioned before please be aware some work had to be done to prepare the sources (like make it independent from our internal infrastructure etc.).

**Comment by papjul** [comment]
Thank you very much, I appreciate!

The licensing issue of the Gradle plugin seems fixed to me.

Should this issue be kept open for the runtime binaries part? There is no licensing issue on that one, as they have always been distributed under a proprietary licence AFAIK, but I know that at least in #1102 people hope you will change your mind as, unfortunately, ObjectBox remains ineligible for inclusion in apps of the default F-Droid repository.

In any case, thank you again for the source code of the plugin part!

**Comment by mnalis** [comment]
Just to clarify as people just quickly scanning the issue still seem to confused (E.g. in  [IzzyOnDroid #884 comment](https://gitlab.com/IzzyOnDroid/repo/-/issues/884#note_2867970098)) by that last comment:

The **ObjectBox is still non-free** software as of end of November 2025, and you still **cannot use ObjectBox in your [FLOSS](https://en.wikipedia.org/wiki/Free_and_open-source_software) apps** (least they _too_ become non-FLOSS). 

---

The issue is tracked in  #1102, but even though the decisions about whether ObjectBox is going to be fully opensourced  were expected "soon" after March of 2024, there still hasn't been any update after year and a half. So, unless  some firm commitments are made by management, it should be assumed that ObjectBox will remain **incompatible** with FOSS apps. 😿
